samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt

```diff
@@ -0,0 +1,514 @@
+                application.getString(R.string.snap_geometry_edits_with_utility_network_rules_app_name)
+    }
+    private val filePath = provisionPath + application.getString(R.string.naperville_geodatabase)
+
+    // Create the mobile map package
+    private val geodatabase = Geodatabase(filePath)
+
+    // Hold references to the subtype sublayers for the distribution and service pipe layers
+    private var distributionPipeLayer: SubtypeSublayer? = null
+    private var servicePipeLayer: SubtypeSublayer? = null
+    private var pipelineLayerName = application.getString(R.string.pipeline_layer_name)
+    private var distributionPipeName = application.getString(R.string.distribution_pipe_name)
+    private var servicePipeLayerName = application.getString(R.string.service_pipe_layer_name)
+
+    // Message dialog view model for handling error messages
+    val messageDialogVM = MessageDialogViewModel()
+
+    // Symbols to help visualize the snap rules behaviors
+    private val rulesPreventSymbol: Symbol =
+        SimpleLineSymbol(SimpleLineSymbolStyle.Solid, Color.red, 4f)
+    private val rulesLimitSymbol: Symbol =
+        SimpleLineSymbol(SimpleLineSymbolStyle.Solid, Color.fromRgba(255, 165, 0 ), 3f)
+    private val noneSymbol: Symbol =
+        SimpleLineSymbol(SimpleLineSymbolStyle.Dash, Color.green, 3f)
+    private val symbols = mutableMapOf(
+        SnapRuleBehavior.RulesPreventSnapping to rulesPreventSymbol,
+        SnapRuleBehavior.RulesLimitSnapping to rulesLimitSymbol,
+        SnapRuleBehavior.None to noneSymbol
+    )
+    private val symbolSwatches = mutableMapOf<SnapRuleBehavior, BitmapDrawable>()
+
+    // Save default renderers to reset the layers when a feature selection is cleared
+    private var defaultDistributionRenderer: Renderer? = null
+    private var defaultServiceRenderer: Renderer? = null
+
+    // Define a geometry editor with a snapping enabled. A reticle tool is ideal for touch devices.
+    val geometryEditor = GeometryEditor().apply {
+        snapSettings.isEnabled = true
+        snapSettings.isFeatureSnappingEnabled = true
+        tool = ReticleVertexTool()
+    }
+
+    // Define a graphics overlay which can also act as a snap source
+    private val defaultGraphicRenderer = SimpleRenderer(SimpleLineSymbol(
+        SimpleLineSymbolStyle.Dash, Color.fromRgba(165, 165, 165), 3f))
+    val graphicsOverlay = GraphicsOverlay(
+        graphics = listOf(Graphic(Geometry.fromJsonOrNull(application.getString(R.string.graphic_geometry_json))))
+    ).apply {
+        id = application.getString(R.string.graphics_overlay_id)
+        renderer = defaultGraphicRenderer
+    }
+
+    // Hold a reference to a selected feature and its related asset group and type
+    private var selectedFeature: ArcGISFeature? = null
+
+    private val _assetGroupNameState = MutableStateFlow("<Nothing selected>")
+    var assetGroupNameState: StateFlow<String> = _assetGroupNameState
```

Comment:
This should be `val`, also I prefer to use `asStateFlow` instead of inline cast here:
```suggestion
    val assetGroupNameState = _assetGroupNameState.asStateFlow()
```